pubtools/pulplib/_impl/fake/match.py

```diff
@@ -96,6 +97,12 @@ def match_field(criteria, obj):
 
 @visit(EqMatcher)
 def match_field_eq(matcher, field, obj):
+    if isinstance(matcher._value, six.string_types):
```

Comment:
A couple of issues with this:

1. I don't think applying a matcher should have the side-effect of mutating it (i.e. this function should not be modifying matcher._value).  If this logic is otherwise correct, I think it should be implemented in a way which doesn't modify any of the passed in objects, like

```
# By default we are using exactly the matcher's stored value
to_match = matcher._value

# But under certain circumstances we might use a different value
if some_conditions(...):
  to_match = something_else(...)

# Now use the value however it's appropriate to do so
if value_matches(to_match):
    ...
```

2. A bigger problem is that the logic just doesn't seem to make sense: it wouldn't be right to assume that, *whenever* a string can be parsed as a datetime, then it's correct to do so.  Strings which look like datetimes could also be stored in actual string fields in Pulp, and then the logic here is going to mess up some matches on those fields.  cdn_published should be an example of this.

Reply:
Actually, cdn_published wasn't a good example as it'll really be stored as a datetime when using fake client.  erratum updated/issued fields are a better example, as they are defined as string timestamps of no strictly enforced format.  So they can happen to contain values matching the logic here and be parsed into timestamps when it's not appropriate to do so.

Reply:
I was trying to match what I saw what the real less_than matcher does compared to the fake matcher. I agree the logic doesn't make sense but I pulled that line of code from the pubtools-pulp code :\

What I've seen is that if you pass datetime it converts the $lt search to YYYY-MM-DDTHH:MM:SSZ but items in search are set to YYYY-MM-DD HH:MM:SS. I can't just change the way the datetime is converted or send a string since the tests expect a datetime. I don't really feel comfortable with changing any core functionality in pulp to fix the issue.

here in target_pulp2.py this is what it sets cdn_published to cdn_published = datetime.datetime.now().strftime('%Y-%m-%dT%H:%M:%S%z')

Reply:
Sorry, I'm still not getting it.

> if you pass datetime it converts the $lt search to YYYY-MM-DDTHH:MM:SSZ

Right, this is $date operator implemented by Pulp, mentioned at https://docs.pulpproject.org/en/2.19/dev-guide/conventions/criteria.html#search-criteria .

> but items in search are set to YYYY-MM-DD HH:MM:SS

What are "items in search" and what do you mean by "set to"?

Could you please reply to https://github.com/release-engineering/pubtools-pulplib/pull/193#pullrequestreview-1048728884 ?